# CSE 163
## Final Project
## Statistical Analysis on Heart Disease Data

Importing basic libraries for analysis:

In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt

Import dataset, and view data:

In [2]:
heart = pd.read_csv('heart_disease_uci.csv')
heart

id  age     sex        dataset               cp  trestbps   chol    fbs  \
0      1   63    Male      Cleveland   typical angina     145.0  233.0   True   
1      2   67    Male      Cleveland     asymptomatic     160.0  286.0  False   
2      3   67    Male      Cleveland     asymptomatic     120.0  229.0  False   
3      4   37    Male      Cleveland      non-anginal     130.0  250.0  False   
4      5   41  Female      Cleveland  atypical angina     130.0  204.0  False   
..   ...  ...     ...            ...              ...       ...    ...    ...   
915  916   54  Female  VA Long Beach     asymptomatic     127.0  333.0   True   
916  917   62    Male  VA Long Beach   typical angina       NaN  139.0  False   
917  918   55    Male  VA Long Beach     asymptomatic     122.0  223.0   True   
918  919   58    Male  VA Long Beach     asymptomatic       NaN  385.0   True   
919  920   62    Male  VA Long Beach  atypical angina     120.0  254.0  False   

              restecg  thalch  exang  oldpeak        slope   ca  \
0      lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1      lv hypertrophy   108.0   True      1.5         flat  3.0   
2      lv hypertrophy   129.0   True      2.6         flat  2.0   
3              normal   187.0  False      3.5  downsloping  0.0   
4      lv hypertrophy   172.0  False      1.4    upsloping  0.0   
..                ...     ...    ...      ...          ...  ...   
915  st-t abnormality   154.0  False      0.0          NaN  NaN   
916  st-t abnormality     NaN    NaN      NaN          NaN  NaN   
917  st-t abnormality   100.0  False      0.0          NaN  NaN   
918    lv hypertrophy     NaN    NaN      NaN          NaN  NaN   
919    lv hypertrophy    93.0   True      0.0          NaN  NaN   

                  thal  num  
0         fixed defect    0  
1               normal    2  
2    reversable defect    1  
3               normal    0  
4               normal    0  
..                 ...  ...  
915                NaN    1  
916                NaN    0  
917       fixed defect    2  
918                NaN    0  
919                NaN    1  

[920 rows x 16 columns]

In [3]:
heart.columns

Index(['id', 'age', 'sex', 'dataset', 'cp', 'trestbps', 'chol', 'fbs',
       'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'],
      dtype='object')

View NA values:

In [4]:
heart.isna().sum()

id            0
age           0
sex           0
dataset       0
cp            0
trestbps     59
chol         30
fbs          90
restecg       2
thalch       55
exang        55
oldpeak      62
slope       309
ca          611
thal        486
num           0
dtype: int64

Here we examind some specific values for thal, which we determined was an import factor in our data:

In [5]:
heart.thal.value_counts()

normal               196
reversable defect    192
fixed defect          46
Name: thal, dtype: int64

In [6]:
heart.shape

(920, 16)

When working with our data, we decided for this statistical analysis to not remove any NA values. We want as much data to go into the below models as possible, and the NA values are not currently affecting our analysis. This just means when considering values such as slope, ca, and thal with lots of NA values that we take their results with a grain of salt.

## Statistical Analysis using the Statsmodels library
### Fitting the entire dataset based on 'num'

To analyze our model, we decided to use the OLS model, or the Ordinary Least Squares model. The OLS model utilizes linear regression to fit the model, and when used with the summary function, we can determine which values are statistically significant in this model. Statistically significant means that the values we get from these variables would be very rare if our null hypothesis was true. Essentially, that they could not just happen randomly so they must be important in our analysis. 

We decided to use the linear regression model, even though our outcomes are categorical. We are using this method because we wanted to analyze which variables are most influential. We are using the values given in the OLS model to determine which are most influential.

From analyzing the dataset, the predictive value is 'num'. Num stands for the predicted stage of heart disease for each patient, and with our analysis we want to see which variables are most influential in predicting these values. In this analysis we leave out location because we want to analyze this variable later.

In [7]:
m = smf.ols("num ~ age + sex + cp + trestbps + chol + fbs + restecg + thalch + exang + oldpeak + slope + ca + thal", data = heart).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.584
Model:                            OLS   Adj. R-squared:                  0.557
Method:                 Least Squares   F-statistic:                     21.82
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           3.38e-43
Time:                        14:45:31   Log-Likelihood:                -354.72
No. Observations:                 299   AIC:                             747.4
Df Residuals:                     280   BIC:                             817.8
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       1.0533      0.750      1.403      0.162      -0.424       2.531
sex[T.Male]                     0.2147      0.118      1.824      0.069      -0.017       0.446
cp[T.atypical angina]          -0.2846      0.157     -1.812      0.071      -0.594       0.025
cp[T.non-anginal]              -0.4625      0.130     -3.551      0.000      -0.719      -0.206
cp[T.typical angina]           -0.6959      0.198     -3.512      0.001      -1.086      -0.306
fbs[T.True]                     0.0164      0.143      0.114      0.909      -0.266       0.299
restecg[T.normal]              -0.1994      0.100     -1.985      0.048      -0.397      -0.002
restecg[T.st-t abnormality]     0.5487      0.428      1.283      0.201      -0.293       1.391
exang[T.True]                   0.1649      0.121      1.360      0.175      -0.074       0.403
slope[T.flat]                   0.0410      0.210      0.195      0.845      -0.372       0.453
slope[T.upsloping]             -0.2248      0.231     -0.972      0.332      -0.680       0.231
thal[T.normal]                 -0.1430      0.224     -0.639      0.523      -0.584       0.297
thal[T.reversable defect]       0.3947      0.213      1.851      0.065      -0.025       0.814
age                            -0.0071      0.006     -1.109      0.268      -0.020       0.006
trestbps                        0.0042      0.003      1.434      0.153      -0.002       0.010
chol                         9.282e-05      0.001      0.096      0.924      -0.002       0.002
thalch                         -0.0044      0.003     -1.619      0.106      -0.010       0.001
oldpeak                         0.1893      0.056      3.395      0.001       0.080       0.299
ca                              0.4200      0.059      7.079      0.000       0.303       0.537
==============================================================================
Omnibus:                       21.441   Durbin-Watson:                   1.910
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               27.359
Skew:                           0.552   Prob(JB):                     1.15e-06
Kurtosis:                       3.988   Cond. No.                     5.22e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.22e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

When we conduct our summary we find 4 significant values, ca, oldpeak, cp(non-aniginal), cp(typical angina). For categorical variables, such as cp, statsmodels automatically makes them into dummy variables because categorical variables would not work in linear regression. 

For our model, we determine statistical significance by looking at p-values and t-values. P-values are probability, which in this summary tells us that probability that this value would occur. We are looking for extremely small values. A p-value of 0.000 means that the probability this value would occur is almost 0%, which tells us it is significant in our model. Additionally, we also need to look at the t-value. The t-value is a very common statistical measure that is created from variability and standard error within the model. For our analysis it is important, the greater the t-value the more evidence there is against the null hypothesis. 

Our 4 values are determined to be significant because their p-values are all very small [0.000 - 0.001]. Additionally, their t-value are all larger than the t-critical value. When analyzing t-values, you should look up the t-critical value on a t-value chart. For our analysis, the t-critical value is 2.101. This is determined by viewing the confidence interval and Df Model values and then matching those values on the t-value table. Our 4 variables are the only ones with t-values greater than the t-critical. This is how we've determined their significance.

Lastly, we can analyze these values by looking at their coefficient (coeff). Firstly we look at the intercept. This tell us num value for our default in our model. The statsmodel library will select the default values automatically, and changing them is very complicated when working with so many variables. Our default values are the categorical variables not listed within the summary table. From the intercept value, the change is the value of each variable. For example, when looking at `oldpeak` we can determine that as `oldpeak` increases by 1 unit, that the predicted level of heart disease will increase by 0.1893. This rise is not as influential as the other variables. The most influential appears to be cp(typical angina). If the patient has typical angina, their stage of heart disease decreases by 0.70, which is almost a whole stage. 

In [8]:
m = smf.ols("num ~ cp + oldpeak + ca", data = heart).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.484
Model:                            OLS   Adj. R-squared:                  0.476
Method:                 Least Squares   F-statistic:                     56.94
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           1.25e-41
Time:                        14:45:32   Log-Likelihood:                -399.30
No. Observations:                 309   AIC:                             810.6
Df Residuals:                     303   BIC:                             833.0
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                 0.6765      0.104      6.505      0.000       0.472       0.881
cp[T.atypical angina]    -0.6569      0.155     -4.237      0.000      -0.962      -0.352
cp[T.non-anginal]        -0.7917      0.125     -6.342      0.000      -1.037      -0.546
cp[T.typical angina]     -0.8635      0.197     -4.389      0.000      -1.251      -0.476
oldpeak                   0.3363      0.048      7.025      0.000       0.242       0.430
ca                        0.4612      0.058      8.016      0.000       0.348       0.574
==============================================================================
Omnibus:                       27.524   Durbin-Watson:                   1.915
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               33.397
Skew:                           0.698   Prob(JB):                     5.60e-08
Kurtosis:                       3.805   Cond. No.                         8.20
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Analyzing by Location
In this analysis we add in the location data with all the variales, and then we analyze with just the most significant values and location.

In [9]:
heart.dataset.value_counts()

Cleveland        304
Hungary          293
VA Long Beach    200
Switzerland      123
Name: dataset, dtype: int64

In [10]:
m = smf.ols("num ~ age + sex + dataset + cp + trestbps + chol + fbs + restecg + thalch + exang + oldpeak + slope + ca + thal", data = heart).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.586
Model:                            OLS   Adj. R-squared:                  0.556
Method:                 Least Squares   F-statistic:                     19.69
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           3.43e-42
Time:                        14:45:32   Log-Likelihood:                -353.85
No. Observations:                 299   AIC:                             749.7
Df Residuals:                     278   BIC:                             827.4
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       1.1167      0.753      1.484      0.139      -0.365       2.598
sex[T.Male]                     0.2171      0.118      1.843      0.066      -0.015       0.449
dataset[T.Hungary]             -0.9031      0.841     -1.074      0.284      -2.558       0.752
dataset[T.Switzerland]      -1.703e-14   1.13e-14     -1.507      0.133   -3.93e-14    5.21e-15
dataset[T.VA Long Beach]       -0.5961      0.841     -0.709      0.479      -2.252       1.060
cp[T.atypical angina]          -0.2867      0.157     -1.823      0.069      -0.596       0.023
cp[T.non-anginal]              -0.4666      0.130     -3.579      0.000      -0.723      -0.210
cp[T.typical angina]           -0.7011      0.198     -3.535      0.000      -1.091      -0.311
fbs[T.True]                     0.0147      0.143      0.103      0.918      -0.268       0.297
restecg[T.normal]              -0.1910      0.101     -1.896      0.059      -0.389       0.007
restecg[T.st-t abnormality]     0.5454      0.428      1.274      0.204      -0.297       1.388
exang[T.True]                   0.1710      0.121      1.408      0.160      -0.068       0.410
slope[T.flat]                   0.0507      0.210      0.242      0.809      -0.362       0.464
slope[T.upsloping]             -0.2187      0.232     -0.944      0.346      -0.675       0.237
thal[T.normal]                 -0.1326      0.224     -0.592      0.555      -0.574       0.309
thal[T.reversable defect]       0.4113      0.214      1.924      0.055      -0.010       0.832
age                            -0.0076      0.006     -1.171      0.243      -0.020       0.005
trestbps                        0.0045      0.003      1.512      0.132      -0.001       0.010
chol                        -1.198e-05      0.001     -0.012      0.990      -0.002       0.002
thalch                         -0.0048      0.003     -1.757      0.080      -0.010       0.001
oldpeak                         0.1883      0.056      3.374      0.001       0.078       0.298
ca                              0.4145      0.060      6.962      0.000       0.297       0.532
==============================================================================
Omnibus:                       20.983   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               27.287
Skew:                           0.534   Prob(JB):                     1.19e-06
Kurtosis:                       4.024   Cond. No.                     4.43e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.61e-28. This might indicate that there are
strong multicollinearity problems

We first analyzed the model by incorporating location into the model to see if those locations are significant in our model. After incorporating them, we have learned than none of the location variables are statistically significant in influencing our stages of heart disease. This does not mean that they do not affect heart disease level, but in this model we cannot eliminate them as random. However, our original 4 most significant variables are still significant, which adds to our confidence in them.

In [11]:
m = smf.ols("num ~ cp + oldpeak + ca + dataset", data = heart).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.488
Model:                            OLS   Adj. R-squared:                  0.474
Method:                 Least Squares   F-statistic:                     35.74
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           1.71e-39
Time:                        14:45:32   Log-Likelihood:                -398.23
No. Observations:                 309   AIC:                             814.5
Df Residuals:                     300   BIC:                             848.1
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
============================================================================================
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept                    0.6911      0.106      6.533      0.000       0.483       0.899
cp[T.atypical angina]       -0.6619      0.156     -4.248      0.000      -0.969      -0.355
cp[T.non-anginal]           -0.8065      0.126     -6.411      0.000      -1.054      -0.559
cp[T.typical angina]        -0.8885      0.198     -4.491      0.000      -1.278      -0.499
dataset[T.Hungary]          -0.3074      0.520     -0.591      0.555      -1.332       0.717
dataset[T.Switzerland]       0.3007      0.408      0.737      0.462      -0.502       1.104
dataset[T.VA Long Beach]    -0.7018      0.637     -1.102      0.271      -1.955       0.551
oldpeak                      0.3404      0.048      7.061      0.000       0.246       0.435
ca                           0.4473      0.059      7.638      0.000       0.332       0.563
==============================================================================
Omnibus:                       26.909   Durbin-Watson:                   1.937
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               32.620
Skew:                           0.685   Prob(JB):                     8.25e-08
Kurtosis:                       3.812   Cond. No.                         24.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

We are now going analyze our location by filtering our data into the US and Europe. To complete our model, we need our observations to be abundant, which is why we were unable to seperate them by individual location. 

In [12]:
us = heart[(heart.dataset == 'VA Long Beach') | (heart.dataset == 'Cleveland')]
europe = heart[(heart.dataset == 'Switzerland') | (heart.dataset == 'Hungary')]

In [13]:
us.shape, europe.shape

((504, 16), (416, 16))

In [14]:
m = smf.ols("num ~ age + sex + cp + trestbps + chol + fbs + restecg + thalch + exang + oldpeak + slope + ca + thal", data = us).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.585
Model:                            OLS   Adj. R-squared:                  0.559
Method:                 Least Squares   F-statistic:                     21.89
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           2.98e-43
Time:                        14:45:32   Log-Likelihood:                -353.44
No. Observations:                 298   AIC:                             744.9
Df Residuals:                     279   BIC:                             815.1
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       1.0953      0.751      1.458      0.146      -0.384       2.574
sex[T.Male]                     0.2177      0.118      1.850      0.065      -0.014       0.449
cp[T.atypical angina]          -0.2855      0.157     -1.817      0.070      -0.595       0.024
cp[T.non-anginal]              -0.4641      0.130     -3.564      0.000      -0.720      -0.208
cp[T.typical angina]           -0.6981      0.198     -3.524      0.000      -1.088      -0.308
fbs[T.True]                     0.0149      0.143      0.104      0.917      -0.267       0.297
restecg[T.normal]              -0.1933      0.101     -1.922      0.056      -0.391       0.005
restecg[T.st-t abnormality]     0.5461      0.428      1.277      0.203      -0.296       1.388
exang[T.True]                   0.1668      0.121      1.376      0.170      -0.072       0.405
slope[T.flat]                   0.0459      0.210      0.219      0.827      -0.367       0.458
slope[T.upsloping]             -0.2212      0.231     -0.956      0.340      -0.677       0.234
thal[T.normal]                 -0.1361      0.224     -0.608      0.544      -0.577       0.304
thal[T.reversable defect]       0.4041      0.213      1.894      0.059      -0.016       0.824
age                            -0.0078      0.006     -1.205      0.229      -0.020       0.005
trestbps                        0.0045      0.003      1.531      0.127      -0.001       0.010
chol                         9.769e-05      0.001      0.101      0.920      -0.002       0.002
thalch                         -0.0048      0.003     -1.748      0.082      -0.010       0.001
oldpeak                         0.1882      0.056      3.376      0.001       0.078       0.298
ca                              0.4171      0.059      7.025      0.000       0.300       0.534
==============================================================================
Omnibus:                       21.197   Durbin-Watson:                   1.913
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               27.282
Skew:                           0.544   Prob(JB):                     1.19e-06
Kurtosis:                       4.006   Cond. No.                     5.22e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.22e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

After conducting our analysis on the US, we started to work on the Europe analysis but we were unable to complete it. We did not have as many observations in the Europe dataframe and we ran into some unexpected error. This is a limitation in our analysis, but through the model fit to the US data we've determined that our 4 variables are still significant even when the data is filtered down to just US data.

### Analyzing by Gender
In this analysis we seperate the data by gender, and in this dataset we only have two values. We then want to examine which are the most significant values for each gender. We're wondering if the statistically significant values change depending on gender, and if it is possible that one gender is skewing the data, especially with the differing number of men and women.

In [15]:
female = heart[heart.sex == 'Female']
male = heart[heart.sex == 'Male']

In [16]:
female.shape, male.shape

((194, 16), (726, 16))

In [17]:
m = smf.ols("num ~ age + sex + dataset + cp + trestbps + chol + fbs + restecg + thalch + exang + oldpeak + slope + ca + thal", data = female).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.789
Model:                            OLS   Adj. R-squared:                  0.743
Method:                 Least Squares   F-statistic:                     17.17
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           1.02e-19
Time:                        14:45:33   Log-Likelihood:                -65.278
No. Observations:                  96   AIC:                             166.6
Df Residuals:                      78   BIC:                             212.7
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       0.1511      1.151      0.131      0.896      -2.140       2.442
dataset[T.Hungary]          -1.507e-14   1.02e-13     -0.147      0.883   -2.19e-13    1.89e-13
dataset[T.Switzerland]          -3e-15   5.17e-15     -0.580      0.563   -1.33e-14    7.29e-15
dataset[T.VA Long Beach]     2.216e-15   1.54e-15      1.439      0.154    -8.5e-16    5.28e-15
cp[T.atypical angina]          -0.3333      0.173     -1.922      0.058      -0.679       0.012
cp[T.non-anginal]              -0.4553      0.151     -3.013      0.003      -0.756      -0.154
cp[T.typical angina]           -0.6486      0.312     -2.079      0.041      -1.270      -0.027
fbs[T.True]                     0.3791      0.192      1.970      0.052      -0.004       0.762
restecg[T.normal]              -0.1163      0.122     -0.957      0.341      -0.358       0.126
restecg[T.st-t abnormality]     0.8197      0.369      2.221      0.029       0.085       1.554
exang[T.True]                  -0.0759      0.166     -0.457      0.649      -0.406       0.255
slope[T.flat]                  -0.1480      0.344     -0.431      0.668      -0.832       0.536
slope[T.upsloping]             -0.3157      0.378     -0.834      0.407      -1.069       0.438
thal[T.normal]                  0.0040      0.615      0.006      0.995      -1.220       1.227
thal[T.reversable defect]       1.1523      0.598      1.927      0.058      -0.038       2.343
age                            -0.0105      0.008     -1.349      0.181      -0.026       0.005
trestbps                        0.0064      0.003      1.849      0.068      -0.000       0.013
chol                           -0.0007      0.001     -0.676      0.501      -0.003       0.001
thalch                          0.0019      0.004      0.512      0.610      -0.005       0.009
oldpeak                         0.0645      0.085      0.755      0.452      -0.106       0.235
ca                              0.4555      0.084      5.406      0.000       0.288       0.623
==============================================================================
Omnibus:                       17.736   Durbin-Watson:                   2.334
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               25.196
Skew:                           0.851   Prob(JB):                     3.38e-06
Kurtosis:                       4.845   Cond. No.                     1.10e+19
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 9.19e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [18]:
m = smf.ols("num ~ age + sex + dataset + cp + trestbps + chol + fbs + restecg + thalch + exang + oldpeak + slope + ca + thal", data = male).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.519
Model:                            OLS   Adj. R-squared:                  0.469
Method:                 Least Squares   F-statistic:                     10.37
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           2.24e-20
Time:                        14:45:33   Log-Likelihood:                -261.36
No. Observations:                 203   AIC:                             562.7
Df Residuals:                     183   BIC:                             629.0
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       1.3544      1.013      1.337      0.183      -0.644       3.353
dataset[T.Hungary]             -1.0090      0.958     -1.053      0.294      -2.899       0.881
dataset[T.Switzerland]       8.277e-15   5.97e-15      1.386      0.168   -3.51e-15    2.01e-14
dataset[T.VA Long Beach]       -0.7691      0.969     -0.794      0.428      -2.680       1.142
cp[T.atypical angina]          -0.3249      0.224     -1.451      0.148      -0.767       0.117
cp[T.non-anginal]              -0.5286      0.186     -2.845      0.005      -0.895      -0.162
cp[T.typical angina]           -0.7574      0.253     -2.994      0.003      -1.257      -0.258
fbs[T.True]                    -0.0631      0.194     -0.325      0.746      -0.446       0.320
restecg[T.normal]              -0.2074      0.140     -1.487      0.139      -0.483       0.068
restecg[T.st-t abnormality]     1.4120      1.043      1.354      0.178      -0.646       3.470
exang[T.True]                   0.1990      0.166      1.196      0.233      -0.129       0.527
slope[T.flat]                   0.1423      0.278      0.513      0.609      -0.405       0.690
slope[T.upsloping]             -0.2499      0.307     -0.815      0.416      -0.855       0.355
thal[T.normal]                  0.0119      0.276      0.043      0.966      -0.532       0.555
thal[T.reversable defect]       0.4367      0.259      1.687      0.093      -0.074       0.948
age                            -0.0044      0.009     -0.492      0.623      -0.022       0.013
trestbps                        0.0043      0.004      0.991      0.323      -0.004       0.013
chol                           -0.0004      0.002     -0.229      0.819      -0.004       0.003
thalch                         -0.0054      0.004     -1.457      0.147      -0.013       0.002
oldpeak                         0.1810      0.075      2.414      0.017       0.033       0.329
ca                              0.3632      0.081      4.471      0.000       0.203       0.523
==============================================================================
Omnibus:                        8.662   Durbin-Watson:                   2.032
Prob(Omnibus):                  0.013   Jarque-Bera (JB):                8.580
Skew:                           0.450   Prob(JB):                       0.0137
Kurtosis:                       3.453   Cond. No.                     8.07e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.14e-29. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

When analyzing by age, we were able to see variation in our influential variables. For women, we found a unique analysis in that they only had 3 siginificant values within their summary table. They kept the `ca` and `cp non-anginal`, but they also are influenced by `restecg[T.st-t abnormality]` which is new. In our previous analysis this value was not significant. For women this value is the most influential. When `restecg[T.st-t abnormality]` is increased by one unit, the level of heart disease is increased by 0.8197 which is almost increases by another stage of heart disease.

For men, we found that the men had the same 4 most significant values. This does not surprise us because men dominate this data set about 7:1, so there is limited information on women.

## Analyzing by Age
In this analysis, we decided to divide the data between less than 55, and 55 and above. We decided to make this distinction due to the size of the data. In order for the model to fit, we needed to have a good number of observations for our dataset. We initially wanted to seperate by decade of age, but that made our dataframes too small and unable to analyze. We reached a value of 55 by finding the median age in the data, which was 54, and rounding it to a nice number.

In [19]:
heart.age.min(), heart.age.max()

(28, 77)

In [20]:
heart.age.median()

54.0

In [21]:
heart.age.value_counts()

54    51
58    43
55    41
56    38
57    38
52    36
62    35
51    35
59    35
53    33
60    32
61    31
48    31
63    30
50    25
41    24
46    24
43    24
64    22
49    22
65    21
44    19
47    19
45    18
42    18
38    16
67    15
39    15
69    13
40    13
66    13
35    11
37    11
68    10
34     7
70     7
74     7
36     6
32     5
71     5
72     4
29     3
75     3
31     2
33     2
76     2
77     2
30     1
28     1
73     1
Name: age, dtype: int64

In [22]:
under55 = heart[heart.age < 55]
above55 = heart[heart.age >= 55]

In [25]:
under55.shape, above55.shape

((472, 16), (448, 16))

In [23]:
m = smf.ols("num ~ age + sex + dataset + cp + trestbps + chol + fbs + restecg + thalch + exang + oldpeak + slope + ca + thal", data = under55).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.648
Model:                            OLS   Adj. R-squared:                  0.595
Method:                 Least Squares   F-statistic:                     12.25
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           1.63e-19
Time:                        14:45:33   Log-Likelihood:                -132.42
No. Observations:                 139   AIC:                             302.8
Df Residuals:                     120   BIC:                             358.6
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       1.8721      1.064      1.760      0.081      -0.234       3.978
sex[T.Male]                     0.1644      0.148      1.112      0.268      -0.128       0.457
dataset[T.Hungary]             -0.9490      0.726     -1.307      0.194      -2.387       0.489
dataset[T.Switzerland]      -9.834e-15   5.39e-15     -1.825      0.071   -2.05e-14    8.36e-16
dataset[T.VA Long Beach]     -1.44e-14   7.12e-15     -2.023      0.045   -2.85e-14   -3.09e-16
cp[T.atypical angina]          -0.3627      0.185     -1.960      0.052      -0.729       0.004
cp[T.non-anginal]              -0.4764      0.163     -2.923      0.004      -0.799      -0.154
cp[T.typical angina]           -0.6543      0.272     -2.402      0.018      -1.194      -0.115
fbs[T.True]                    -0.3338      0.204     -1.636      0.105      -0.738       0.070
restecg[T.normal]              -0.1137      0.127     -0.894      0.373      -0.366       0.138
restecg[T.st-t abnormality] -2.323e-16   3.52e-16     -0.660      0.511   -9.29e-16    4.65e-16
exang[T.True]                   0.0937      0.167      0.560      0.577      -0.238       0.425
slope[T.flat]                   0.4416      0.289      1.529      0.129      -0.130       1.014
slope[T.upsloping]              0.0881      0.296      0.297      0.767      -0.498       0.674
thal[T.normal]                 -0.3918      0.383     -1.024      0.308      -1.149       0.366
thal[T.reversable defect]       0.3672      0.370      0.991      0.323      -0.366       1.101
age                            -0.0138      0.012     -1.150      0.253      -0.038       0.010
trestbps                       -0.0012      0.004     -0.283      0.778      -0.010       0.007
chol                           -0.0003      0.001     -0.194      0.847      -0.003       0.003
thalch                         -0.0035      0.004     -0.899      0.370      -0.011       0.004
oldpeak                         0.2420      0.077      3.133      0.002       0.089       0.395
ca                              0.3394      0.084      4.046      0.000       0.173       0.506
==============================================================================
Omnibus:                       17.521   Durbin-Watson:                   1.892
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               28.345
Skew:                           0.622   Prob(JB):                     7.00e-07
Kurtosis:                       4.829   Cond. No.                     1.42e+22
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 6.94e-38. This might indicate that there are
strong multicollinearity problems

In [24]:
m = smf.ols("num ~ age + sex + dataset + cp + trestbps + chol + fbs + restecg + thalch + exang + oldpeak + slope + ca + thal", data = above55).fit()
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    num   R-squared:                       0.556
Model:                            OLS   Adj. R-squared:                  0.496
Method:                 Least Squares   F-statistic:                     9.235
Date:                Mon, 13 Mar 2023   Prob (F-statistic):           1.06e-16
Time:                        14:45:34   Log-Likelihood:                -202.55
No. Observations:                 160   AIC:                             445.1
Df Residuals:                     140   BIC:                             506.6
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       1.1936      1.530      0.780      0.437      -1.831       4.218
sex[T.Male]                     0.3514      0.186      1.888      0.061      -0.017       0.719
dataset[T.Hungary]             4.2e-15    3.7e-15      1.135      0.258   -3.12e-15    1.15e-14
dataset[T.Switzerland]       1.293e-14   1.41e-14      0.915      0.362    -1.5e-14    4.09e-14
dataset[T.VA Long Beach]       -0.3261      0.958     -0.340      0.734      -2.220       1.568
cp[T.atypical angina]          -0.2937      0.258     -1.138      0.257      -0.804       0.217
cp[T.non-anginal]              -0.3180      0.216     -1.476      0.142      -0.744       0.108
cp[T.typical angina]           -0.8758      0.294     -2.974      0.003      -1.458      -0.294
fbs[T.True]                     0.1707      0.209      0.816      0.416      -0.243       0.584
restecg[T.normal]              -0.3139      0.159     -1.970      0.051      -0.629       0.001
restecg[T.st-t abnormality]     0.5879      0.492      1.196      0.234      -0.384       1.560
exang[T.True]                   0.2275      0.177      1.283      0.202      -0.123       0.578
slope[T.flat]                  -0.3934      0.318     -1.236      0.218      -1.023       0.236
slope[T.upsloping]             -0.5849      0.368     -1.588      0.115      -1.313       0.143
thal[T.normal]                  0.1751      0.307      0.571      0.569      -0.431       0.781
thal[T.reversable defect]       0.5238      0.280      1.868      0.064      -0.030       1.078
age                            -0.0150      0.018     -0.844      0.400      -0.050       0.020
trestbps                        0.0072      0.004      1.700      0.091      -0.001       0.016
chol                         -5.64e-05      0.001     -0.041      0.967      -0.003       0.003
thalch                         -0.0037      0.004     -0.929      0.354      -0.012       0.004
oldpeak                         0.1115      0.083      1.341      0.182      -0.053       0.276
ca                              0.5074      0.089      5.683      0.000       0.331       0.684
==============================================================================
Omnibus:                        4.196   Durbin-Watson:                   1.814
Prob(Omnibus):                  0.123   Jarque-Bera (JB):                3.797
Skew:                           0.367   Prob(JB):                        0.150
Kurtosis:                       3.177   Cond. No.                     4.09e+18
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.05e-30. This might indicate that there are
strong multicollinearity problems

For our age analysis, we found that patients under 55 exhibit the same 4 most statistically significant values when we conduct our OLS model summary. However, for patients 55 and above, they only have 2 significant values `ca` and `cp[T.typical angina]`. The other two values we've been seeing in other models cannot be ruled out as random in our 55 and above summary table.